### PR TITLE
jsonify: add jump to definition and `any` support

### PR DESCRIPTION
**Pull request by sachinraja**
These changes are upstreamed from [work I did for tRPC](https://github.com/trpc/trpc/pull/3261) and [Remix](https://github.com/remix-run/remix/pull/4689).

For data inferred from `Jsonify`, properties do not have jump to definition. This is because key remapping which breaks jump to definition support for TS (see https://github.com/microsoft/TypeScript/issues/47813). I changed it to use a different strategy for filtering that keeps jump to definition.

I also added support for the `any` type, previously `Jsonify<any>` was `never` and other variations of that were incorrect too.

I refactored the implementation because the previous one did not flow well with these changes.

cc @MichaelDeBoey

### CODE REVIEW

**Review by MichaelDeBoey — APPROVED** [review]
Thanks @sachinraja, this is **_really_** awesome! 🤩

Once this one's merged, we can start using it in the Remix codebase 😍

File: source/jsonify.d.ts [hunk 1]
```diff
@@ -16,17 +36,53 @@ type BaseKeyFilter<Type, Key extends keyof Type> = Key extends symbol
...
-// Returns never if the key or property is not jsonable or optional otherwise return the key.
-type RequiredKeyFilter<Type, Key extends keyof Type> = undefined extends Type[Key]
-	? never
-	: BaseKeyFilter<Type, Key>;
+/**
+ * Returns the required keys.
+ */
+type FilterDefinedKeys<TObject extends object> = Exclude<
+{
+	[TKey in keyof TObject]: IsAny<TObject[TKey]> extends true
```
Comment (MichaelDeBoey) [inline source/jsonify.d.ts:44]:
```suggestion
	[Key in keyof TObject]: IsAny<TObject[Key]> extends true
```

File: source/jsonify.d.ts [hunk 2]
```diff
@@ -16,17 +36,53 @@ type BaseKeyFilter<Type, Key extends keyof Type> = Key extends symbol
...
+		? TKey
+		: undefined extends TObject[TKey]
```
Comment (MichaelDeBoey) [inline source/jsonify.d.ts:46]:
```suggestion
		: undefined extends TObject[Key]
```

File: source/jsonify.d.ts [hunk 3]
```diff
@@ -16,17 +36,53 @@ type BaseKeyFilter<Type, Key extends keyof Type> = Key extends symbol
...
+			? never
+			: TObject[TKey] extends undefined
```
Comment (MichaelDeBoey) [inline source/jsonify.d.ts:48]:
```suggestion
			: TObject[Key] extends undefined
```

File: source/jsonify.d.ts [hunk 4]
```diff
@@ -16,17 +36,53 @@ type BaseKeyFilter<Type, Key extends keyof Type> = Key extends symbol
...
+				? never
+				: BaseKeyFilter<TObject, TKey>;
```
Comment (MichaelDeBoey) [inline source/jsonify.d.ts:50]:
```suggestion
				: BaseKeyFilter<TObject, Key>;
```

File: source/jsonify.d.ts [hunk 5]
```diff
@@ -16,17 +36,53 @@ type BaseKeyFilter<Type, Key extends keyof Type> = Key extends symbol
...
 
-// Returns never if the key or property is not jsonable or required otherwise return the key.
-type OptionalKeyFilter<Type, Key extends keyof Type> = undefined extends Type[Key]
-	? Type[Key] extends undefined
+/**
+ * Returns the optional keys.
+ */
+type FilterOptionalKeys<TObject extends object> = Exclude<
+{
+	[TKey in keyof TObject]: IsAny<TObject[TKey]> extends true
```
Comment (MichaelDeBoey) [inline source/jsonify.d.ts:60]:
```suggestion
	[Key in keyof TObject]: IsAny<TObject[Key]> extends true
```

File: source/jsonify.d.ts [hunk 6]
```diff
@@ -16,17 +36,53 @@ type BaseKeyFilter<Type, Key extends keyof Type> = Key extends symbol
...
-type RequiredKeyFilter<Type, Key extends keyof Type> = undefined extends Type[Key]
-	? never
-	: BaseKeyFilter<Type, Key>;
+/**
+ * Returns the required keys.
+ */
+type FilterDefinedKeys<TObject extends object> = Exclude<
+{
+	[TKey in keyof TObject]: IsAny<TObject[TKey]> extends true
+		? TKey
```
Comment (MichaelDeBoey) [inline source/jsonify.d.ts:45]:
```suggestion
		? Key
```

File: source/jsonify.d.ts [hunk 7]
```diff
@@ -16,17 +36,53 @@ type BaseKeyFilter<Type, Key extends keyof Type> = Key extends symbol
...
+/**
+ * Returns the optional keys.
+ */
+type FilterOptionalKeys<TObject extends object> = Exclude<
+{
+	[TKey in keyof TObject]: IsAny<TObject[TKey]> extends true
 		? never
-		: BaseKeyFilter<Type, Key>
-	: never;
+		: undefined extends TObject[TKey]
```
Comment (MichaelDeBoey) [inline source/jsonify.d.ts:62]:
```suggestion
		: undefined extends TObject[Key]
```

File: source/jsonify.d.ts [hunk 8]
```diff
@@ -16,17 +36,53 @@ type BaseKeyFilter<Type, Key extends keyof Type> = Key extends symbol
...
+			? TObject[TKey] extends undefined
```
Comment (MichaelDeBoey) [inline source/jsonify.d.ts:63]:
```suggestion
			? TObject[Key] extends undefined
```

File: source/jsonify.d.ts [hunk 9]
```diff
@@ -16,17 +36,53 @@ type BaseKeyFilter<Type, Key extends keyof Type> = Key extends symbol
...
+				? never
+				: BaseKeyFilter<TObject, TKey>
```
Comment (MichaelDeBoey) [inline source/jsonify.d.ts:65]:
```suggestion
				: BaseKeyFilter<TObject, Key>
```

**Review by MichaelDeBoey — COMMENTED** [review]
Forgot to mention these 🙈

No problem if you decide to not change these though 🤷‍♂️

File: source/jsonify.d.ts [hunk 10]
```diff
@@ -1,12 +1,32 @@
...
 // Note: The return value has to be `any` and not `unknown` so it can match `void`.
 type NotJsonable = ((...args: any[]) => any) | undefined | symbol;
 
+/** JSON serialize [tuples](https://www.typescriptlang.org/docs/handbook/2/objects.html#tuple-types) */
+type JsonifyTuple<T extends [unknown, ...unknown[]]> = {
+	[k in keyof T]: T[k] extends NotJsonable ? null : Jsonify<T[k]>;
+};
+
+type FilterJsonableKeys<T extends object> = {
+	[Key in keyof T]: T[Key] extends NotJsonable ? never : Key;
```
[the remark this answers is not in the document]
Comment (sindresorhus) [reply source/jsonify.d.ts:22]: I prefer `Key` too.

File: source/jsonify.d.ts [hunk 11]
```diff
@@ -1,12 +1,32 @@
...
-import type {Merge} from './merge';
 import type {NegativeInfinity, PositiveInfinity} from './numeric';
+import type {Simplify} from './simplify';
 import type {TypedArray} from './typed-array';
 
+/*
+ * `any` is the only type that can let you equate `0` with `1`
+ * See https://stackoverflow.com/a/49928360/1490091
+ */
+type IsAny<T> = 0 extends 1 & T ? true : false;
```
Comment (sindresorhus) [inline source/jsonify.d.ts:11]: Can you move this to internal.d.ts? We keep reusable types there.

File: source/jsonify.d.ts [hunk 12]
```diff
@@ -1,12 +1,32 @@
 import type {JsonPrimitive, JsonValue} from './basic';
 import type {EmptyObject} from './empty-object';
-import type {Merge} from './merge';
 import type {NegativeInfinity, PositiveInfinity} from './numeric';
+import type {Simplify} from './simplify';
 import type {TypedArray} from './typed-array';
 
+/*
+ * `any` is the only type that can let you equate `0` with `1`
+ * See https://stackoverflow.com/a/49928360/1490091
```
Comment (sindresorhus) [inline source/jsonify.d.ts:9]: Don't prefix comment lines with `*`.

File: source/jsonify.d.ts [hunk 13]
```diff
@@ -16,17 +36,53 @@ type BaseKeyFilter<Type, Key extends keyof Type> = Key extends symbol
...
+				? never
+				: BaseKeyFilter<TObject, TKey>
+			: never;
+}[keyof TObject],
+undefined
+>;
+
+/*
+ * For an object T, if it has any properties that are a union with `undefined`,
+ * make those into optional properties instead.
```
Comment (sindresorhus) [inline source/jsonify.d.ts:73]: Don't hard-wrap.

File: source/jsonify.d.ts [hunk 14]
```diff
@@ -16,17 +36,53 @@ type BaseKeyFilter<Type, Key extends keyof Type> = Key extends symbol
...
+ *
+ * Example: { a: string | undefined} --> { a?: string}
```
Comment (sindresorhus) [inline source/jsonify.d.ts:75]: Use the same example formatting we use in other types.

File: source/jsonify.d.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (sachinraja) [reply source/jsonify.d.ts:22]: Fixed

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: source/jsonify.d.ts (hunk 11 above)
Comment (sachinraja) [reply source/jsonify.d.ts]: done

**Review by MichaelDeBoey — APPROVED** [review]

**Comment by sindresorhus** [comment]
Nice work 👍

**Pull request merged**